trufflescripts/watch_events.js

```diff
@@ -0,0 +1,155 @@
+  const [master, seller, buyer] = web3.eth.accounts
+
+  const eth = await TokenETH.deployed()
+  const gno = await TokenGNO.deployed()
+  const dx = await DutchExchangeETHGNO.deployed()
+
+  const addr2acc = {
+    [master]: 'Master',
+    [seller]: 'Seller',
+    [buyer]: 'Buyer',
+    [dx.address]: 'DutchExchangeETHGNO',
+    [eth.address]: 'ETH',
+    [gno.address]: 'GNO',
+  }
+
+  const printTime = () => (!argv.log ? `${new Date().toLocaleTimeString()}\t` : '')
+
+  // when logging, read from the very first block
+  const filterObj = argv.log && { fromBlock: 0 }
+  // and wait  for all events before printing them as they will be out of order
+  const logPromises = argv.log && []
+
+  const printLog = (name, err, log) => {
+    if (err) {
+      console.error(err)
+      return
+    }
+
+    const {
+      args,
+      event,
+      logIndex,
+      transactionIndex,
+      blockNumber,
+    } = log
+
+    const addrRegex = /^0x\w{40}$/
+    for (const arg of Object.keys(args)) {
+      const val = args[arg]
+
+      if (val.toNumber) {
+        // convert BigNumbers
+        args[arg] = val.toNumber()
+        // if address
+      } else if (typeof val === 'string' && addrRegex.test(val)) {
+        args[arg] = addr2acc[val] || val
+      }
+    }
+
+    const verbose = argv.v ? `${blockNumber}/${transactionIndex}/${logIndex}\t` : ''
+
+    console.log(`${printTime()}${verbose}${name}::${event}`, args)
+  }
+
+  const printLogs = (logs) => {
+    // logs will be sorted by event type and then by time
+    // need to manually sort by time
+    logs.sort((a, b) => {
+      // first sort by blockNumber
+      let diff = a.blockNumber - b.blockNumber
+      if (diff) return diff
+
+      // if blockNumbers are the same,
+      // sort by transactionIndex
+      diff = a.transactionIndex - b.transactionIndex
+      if (diff) return diff
+
+      // finally by logIndex
+      return a.logIndex - b.logIndex
+    }).forEach(log => printLog(log.contractName, null, log))
+  }
+
+  const watch = (name, contract, events) => {
+    const processEvent = (event) => {
+      if (argv.log) {
+        // if logging, then first wait to gather all events
+        logPromises.push(new Promise((res, rej) => event.get((err, logs) => {
+          if (err) {
+            if (err.message === 'contract[e] is not a function') {
+              console.warn(`contract ${name} doesn't have ${event} event`)
+            } else return rej(err)
+          }
+
+          // additionally assign to each event its contract name
+          return res(logs.map(log => Object.assign({ contractName: name }, log)))
+        })))
+      } else event.watch(printLog.bind(null, name))
+    }
+
+    const eventsArray = typeof events === 'string' ? events.split(',') : ['allEvents']
+
+    eventsArray.forEach((e) => {
+      try {
+        processEvent(contract[e](filterObj))
+      } catch (error) {
+        if (error.message === 'contract[e] is not a function') {
```

Comment:
haven't tested but should be `if (error.message === `${contract[e]} is not a function`) {...}` - no?

Reply:
No, that's the actual message, even if it isn't helpful at all. Stack trace works in mysterious ways.